### PR TITLE
Add SciPy Global Optimization software

### LINKED ISSUE #822: Add SciPy Global Optimization fitting software
Add the differential_evolution, shgo and dual_annealing global optimization solvers from SciPy.

### PR DESCRIPTION
**Pull request by jfowkes**
#### Description of Work

Closes #822 

#### Testing Instructions

1. Check that all three SciPy GO minimisers run on a simple test problem with bounds
2. Check that docs have been updated to include scipy_go
3. Check that the new global_optimization algorithm type label works correctly
4. Check that the new MissingBoundsError is correctly raised

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?

### CODE REVIEW

**Comment by jfowkes** [comment]
@jess-farmer any idea why the default regression tests are not running? I'm getting an error deep in fitbenchmarking: 
```
E       UnboundLocalError: local variable 'unselected_minimzers' referenced before assignment

fitbenchmarking/core/fitting_benchmarking.py:129: UnboundLocalError
```

**Review by jess-farmer — CHANGES_REQUESTED** [review]
Looks good! I just have a few comments:

- There are a few pylint warnings for the scipy_go_controller:
 ```
 fitbenchmarking/controllers/scipy_go_controller.py:59:8: W0201: Attribute '_maxiter' defined outside __init__ (attribute-defined-outside-init)
fitbenchmarking/controllers/scipy_go_controller.py:79:8: W0201: Attribute 'result' defined outside __init__ (attribute-defined-outside-init)
fitbenchmarking/controllers/scipy_go_controller.py:82:8: W0201: Attribute '_status' defined outside __init__ (attribute-defined-outside-init)
```
- I am finding that `differential_evolution` has an error flag 3 every time I run fitbenchmarking - printing `self.result` shows that although the solver successfully converged there isn't `status` attribute in the results object and this seems to be what's causing the problem.

File: fitbenchmarking/controllers/scipy_go_controller.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            kwargs = {"maxiter": self._maxiter}
+        elif self.minimizer == "shgo":
+            kwargs = {"options": {"maxiter": self._maxiter,
+                                  "jac": self.jacobian.eval_cost}}
+        elif self.minimizer == "dual_annealing":
+            kwargs = {"maxiter": self._maxiter, "local_search_options": {
+                "jac": self.jacobian.eval_cost}}
+        fun = self.cost_func.eval_cost
+        if self.value_ranges is None:
+            raise MissingBoundsError("SciPy GO requires bounds on parameters")
```
[the remark this answers is not in the document]
Comment (jfowkes) [reply fitbenchmarking/controllers/scipy_go_controller.py:75]: Agreed will move this into `setup`, also good catch on the `inf` values in `self.value_ranges` will check for that aswell.

File: fitbenchmarking/controllers/tests/test_controllers.py [hunk 2]
```diff
@@ -771,6 +777,40 @@ def test_matlab(self):
             self.shared_tests.check_diverged(controller)
 
 
+@run_for_test_types(TEST_TYPE, 'default', 'all')
```
[the remark this answers is not in the document]
Comment (jfowkes) [reply fitbenchmarking/controllers/tests/test_controllers.py:780]: Ah I was confused as to why it wasn't finding mantid, thanks for the fix!

**Comment by jfowkes** [comment]
> Looks good! I just have a few comments:
> 
> * There are a few pylint warnings for the scipy_go_controller:

Good spot, will fix those!

> * I am finding that `differential_evolution` has an error flag 3 every time I run fitbenchmarking - printing `self.result` shows that although the solver successfully converged there isn't `status` attribute in the results object and this seems to be what's causing the problem.

Yeah this is issue #826 cropping up again, I'll implement per-solver flag handling (it's not going to be pretty as for some solvers I'll have to parse the return messages but it should work consistently at least).

**Comment by jess-farmer** [comment]
@jfowkes Sorry to add another comment, but just wanted to check if you are able to get results for `shgo`; I've just been checking through your changes and realised that even after leaving fitbenchmarking running for an hour it's still stuck on the `shgo` minimizer - not sure if it's just my computer or something that needs looking into!

**Comment by jfowkes** [comment]
@jess-farmer good question! The problem is that these SciPy global optimization routines are really not very good and I have them set to 1,000 iterations by default. Perhaps we should reduce the iteration limit for `shgo` to 100 and hope?

**Comment by jess-farmer** [comment]
@jfowkes yeah that sounds like a good idea to reduce the iteration limit, it would probably also be worth adding a note to the docs (maybe under known issues?) mentioning that `shgo` is particularly slow running and is probably best avoided!

**Comment by jfowkes** [comment]
@jess-farmer good idea, have done!

**Comment by jess-farmer** [comment]
Great, thanks, I'll now approve and merge!

**Pull request merged**